feat: add support outputting rekor response on signing

vishal-chdhry opened the pull request:
<!--
Thanks for opening a pull request! Please do not just delete this text.  The three fields below are mandatory.

Please remember to:
- This PR requires an issue. If it is a new feature, the issue should proceed the PR and will have allowed sufficent time for discussions to take place. Please use
issue tags such as "Closes #XYZ" or "Resolves sigstore/repo-name#XYZ".
  [documentation](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)
- ensure your commits are signed-off, as sigstore uses the [DCO](https://en.wikipedia.org/wiki/Developer_Certificate_of_Origin) using `git commit -s`, or `git commit -s --amend` if you want to amend already existing commits
- lastly, ensure there are no merge commits!

Thank you :)
-->

#### Summary
<!--
 Explain the **motivation** for making this change. What existing problem does the pull request solve? How can reviewers test this PR?
-->

Closes: #3110

Previously, there's no support for outputting a bundle with `cosign sign` ([code](https://github.com/sigstore/cosign/blob/main/cmd/cosign/cli/sign/sign.go#L266-L298)), only `sign-blob` ([code](https://github.com/sigstore/cosign/blob/main/cmd/cosign/cli/sign/sign_blob.go#L135-L152)). This PR adds support for outputting rekor response on signing. This supports use-cases where a signer does not want to attach metadata immediately to the container.

#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates)
* API additions—new endpoint, new response fields, or newly accepted request parameters
* Database changes (any)
* Websocket additions or changes
* Anything noteworthy to an administrator running private sigstore instances (err on the side of over-communicating)
* New features and improvements, including behavioural changes, UI changes and CLI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense.

-->
Added support outputting rekor response on signing with `cosign sign`
#### Documentation
<!--

Does this change require an update to documentation? How will users implement your new feature?

Please reference a PR within https://docs.sigstore.dev

-->

haydentherapper commented on cmd/cosign/cli/sign/sign.go:
```diff
@@ -591,3 +617,33 @@ func (c *SignerVerifier) Bytes(ctx context.Context) ([]byte, error) {
 	return pemBytes, nil
 }
+
+func fetchSignedPayload(sig oci.Signature) (*cosign.SignedPayload, error) {
```
Can this return a `LocalSignedPayload` instead to be in sync with sign-blob? It should be loadable with https://github.com/sigstore/cosign/blob/main/pkg/cosign/fetch.go#L199.

haydentherapper reviewed (changes requested):
Can you add a test for this too?

vishal-chdhry commented:
@haydentherapper I have added a test and updated the logic, can you please review it

Besides, I cannot run/debug tests in `test/e2e_test.go`. I think `gopls` is not working for me in that file, autocomplete does not work either. Is there anything I need to configure? I am using VSCode.

haydentherapper commented on test/e2e_test.go:
```diff
@@ -951,6 +951,51 @@ func TestRekorBundle(t *testing.T) {
 
+func TestRekorOutput(t *testing.T) {
+	// turn on the tlog
+	defer setenv(t, env.VariableExperimental.String(), "1")()
```
Can you remove, the experimental flag is no longer used?

haydentherapper reviewed (dismissed):
One tiny comment, looks good!

vishal-chdhry commented:
@haydentherapper Removed the flag
